numpy/core/_internal.py

```diff
@@ -317,7 +316,9 @@ def data_as(self, obj):
 
         The returned pointer will keep a reference to the array.
         """
-        return self._ctypes.cast(self._data, obj)
+        ptr = self._ctypes.cast(self._data, obj)
+        ptr._arr = self._arr
```

Comment:
Can you do this on line 298 instead?

Reply:
No. If _arr is assigned to ._data at L298, ctypes.cast will create a reference to ._data in ._data._objects when calling .ctypes.data_as(...) . This causes the circular reference to ._data. All attributes of ._data cannot be released automatically. ._arr has to be assigned lazily.

Reply:
So to be clear, even if we take this patch as is, `arr.ctypes.data_as(c_void_p).cast(something_else)` creates a circular reference?

I'm wondering whether it would make more sense to focus on an upstream patch.

Reply:
Yes, call .cast() again creates another circular reference.

I doubt whether this bug of CPython can be fixed soon. https://bugs.python.org/issue12836. It was noticed in 2011. Except for a few discussion around 2013, no updates since then.